### PR TITLE
Metacal script

### LINKED ISSUE #3: : Setup automated tool to evaluate shear bias and compare against metacal equivalent

### PR DESCRIPTION
**Pull request by b-remy**

### CODE REVIEW

**Comment by b-remy** [comment]
So far the Metacal result with a constant shear on 1024 images is the following

```
1024/1024 100%

sigma_n: 5e-06
Metacalibration
----------------------
S/N: 31081.3
R11: 0.00769911
shear [0.05852934 0.02753376]
shear true [0.05, 0.0]
m: 0.170587 +/- 0.383094 (99.7% conf)
c: 0.0275338 +/- 0.0409228 (99.7% conf)
```

I remark that even with a high SNR the residual bias is very high, compared to the ngmix metacal example with a similar SNR. @EiffL do you have any idea what could be wrong?

File: metacal.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    noise = sigma_n
+    wt = im*0 + 1.0/noise**2
+
+    psf_im = psf.drawImage(scale=scale).array
+
+    psf_obs = ngmix.Observation(
+        psf_im,
+    )
+
+    obs = ngmix.Observation(
```
Comment (EiffL) [inline metacal.py:241]: @b-remy When you create obs you should specify the pixel scale, I think this might be one of the sources of the issues.

see here: https://github.com/CosmoStat/autometacal/blob/f5958c41c091b229017be5f072f1c58b94120d80/examples/metacal.py#L296

You can probably remove the mis-centering question.

**Comment by EiffL** [comment]
I added a comment, I'm very suspicious of this R11 value: 
R11: 0.00769911

it would mean that your have almost no response to shear

File: metacal.py (hunk 1 above)
Comment (b-remy) [reply metacal.py:253]: Exact. I added psf and image jacobians in my code. But I don't get to which transformation these jacobians correspond. Do you have an idea?

**Comment by EiffL** [comment]
So the weight_fwhm defines the size of the gaussian window used to measure weighted moments. It is expressed in arcsecs. 
Based on you pixel scale and roughly on the size you expect for your galaxies, you can figure out what to use.

The window should be almost zero at the edges of the stamps

File: metacal.py (hunk 1 above)
Comment (EiffL) [reply metacal.py:253]: When you only specify the pixel scale, the Jacobian is purely diagonal. No transform is applied to the image, it simply indicates how you go from pixel units to arcsecs

**Comment by b-remy** [comment]
Actually the comparison to the ngmix metacal example is not completely fair, since the latter assume fixed size Exponential profiles with no intrinsic ellipticity.

So I made a new toy model ([`toymodel0.py`](toymodel0.py)) to have the exact same data as in [`ngmix.examples.metacal.metacal.py`](https://github.com/esheldon/ngmix/blob/master/examples/metacal/metacal.py), i.e. assuming:
- fixed size `hlr=0.5`
- Moffat psf
- no intrinsic ellipticity
- `shear = [0.01, 0.]`
- `noise = 1e-6`
- no shift

The [`metacal0.py`](https://github.com/b-remy/gems/blob/metacal/metacal0.py) script reproduces exactly [`ngmix.examples.metacal.metacal.py`](https://github.com/esheldon/ngmix/blob/master/examples/metacal/metacal.py), except that the galaxy images are created from our simulations.

Both scripts yield different results, `m: 0.100325` vs `m: 0.000185307 `
- `metacal0.py`:
```
S/N: 79531.4
R11: 0.356769
m: 0.100325 +/- 0.000462569 (99.7% conf)
c: 0.0142586 +/- 4.66907e-06 (99.7% conf)
```
- `ngmix.examples.metacal.metacal.py`:
```
S/N: 80113.7
R11: 0.35164
m: 0.000185307 +/- 0.000401216 (99.7% conf)
c: 9.56549e-07 +/- 3.8409e-06 (99.7% conf)
```

If you want to reproduce these results you can:
1. create simulations `python toymodel0.py --save=True --plot=False --N=32`
2. run metacal with our sims `python metacal0.py --filename=data/sims_toymodel0_0.fits`
3. run metacal with galsim sims `ngmix.examples.metacal.metacal.py` (with no shift, i.e. `dx = dy = 0.`)

So assuming that I haven't forgotten anything, the metacalibration is the same. Only the codes to create images differ.

Maybe the problem is coming from GalFlow?

In GalFlow we have tests on light profile generation. The two things I am currently suspecting are:
- [1] the way we convolve with the PSF in GalFlow
- [2] the shear tranform in GalFlow. We validated this transformation with a pixel residual test, but I think we have also noticed a little bias when averaging the shape measurements after shearing in our automecatal project.

In order to check [1], I made another script [`ngmix/metacal.py`](https://github.com/b-remy/gems/blob/metacal/ngmix/metacal.py), which also does the same thing as `ngmix.examples.metacal.metacal.py`, expect that we can choose to perform the PSF convolution with GalFlow (`--galflowconv=1`) or galsim (`--galflowconv=0`).

- Galsim convolution (regular ngmix example)
```
S/N: 80113.7
R11: 0.35164
m: 0.000185307 +/- 0.000401216 (99.7% conf)
c: 9.56549e-07 +/- 3.8409e-06 (99.7% conf)
```

- GalFlow convolution
```
S/N: 79469.3
R11: 0.359314
m: -0.0460361 +/- 0.000391725 (99.7% conf)
c: 0.0140135 +/- 3.96343e-06 (99.7% conf)
```

So the way we perform the convolution also seems to be a source of error.

**Comment by EiffL** [comment]
hmmmm, ok, yeah I think it's  a problem of postage stamp size, we know that the convolution/shear only works for odd size images in autometacal. This is related to https://github.com/DifferentiableUniverseInitiative/GalFlow/issues/8 

I'm not sure the psf convolution using galflow.convolve here is working correctly

**Comment by EiffL** [comment]
So, I think the thing that doesnt work is when the input image is in the pixels domain. The kconvolve function should work as long as both image and PSF are already provided in kspace.

**Comment by EiffL** [comment]
I've proposed a fix to the `ngmix/metacal.py` example, to do convolutions in kspace, and using odd number of pixels to avoid half pixel shifts compared to galsim

Here are the results of that script using tensorflow convolutions:
```python
S/N: 80162.7
R11: 0.351112
m: -0.000152554 +/- 0.000409342 (99.7% conf)
c: 1.00323e-06 +/- 4.13362e-06 (99.7% conf)
```

**Comment by b-remy** [comment]
Last commit [4d3d23d](https://github.com/b-remy/gems/commit/4d3d23d066dd9dbb932dda998728a031e69966ec) validates that the convolution with GalFlow is working well, as long as we consider **even stamp sizes**.

With GalFlow:
```
S/N: 84713.4
R11: 0.401513
m: 7.92014e-06 +/- 0.000347655 (99.7% conf)
c: -9.48739e-08 +/- 3.49117e-06 (99.7% conf)
```

With Galsim
```
S/N: 84713.9
R11: 0.401513
m: 1.01552e-05 +/- 0.000347654 (99.7% conf)
c: -4.69012e-08 +/- 3.49115e-06 (99.7% conf)
```
The difference with your fix @EiffL is that I now have a real image of galaxy as input of the convolution.

Now I am going to try with `toymodel0` and test the shear.

**Comment by b-remy** [comment]
Note that the same computation is worse in GalFlow when using an Exponential light profile instead of a Gaussian:
GalFlow:
```
S/N: 79324.2
R11: 0.354662
m: -0.00676094 +/- 0.000411014 (99.7% conf)
c: -1.03561e-07 +/- 4.19112e-06 (99.7% conf)
```
Galsim
```
S/N: 80182.9
R11: 0.351485
m: -1.51384e-05 +/- 0.000412506 (99.7% conf)
c: -4.61679e-08 +/- 4.20315e-06 (99.7% conf)
```

**Comment by EiffL** [comment]
Nice !!! Yeah it's possible that performance degrades with exponential. I
guess the profile is more peaky, which means you need to sample on a finer
grid to do the deconvolution.

**Comment by b-remy** [comment]
Now that we have evaluated the impact of the convolution, let's check the impact of the shear.

`toymodel0.py` generates Gaussian light profiles (as I did in `ngmix/metacal.py`, for fair comparison), and performs the convolution with the PSF with the validated tensorflow functions for odd stamp sizes. The `metacal0.py` script then returns:
```
sigma_n: 1e-06
Metacalibration
----------------------
S/N: 84834.2
R11: 0.40136
shear [ 1.0685088e-02 -1.6877929e-06]
shear true [0.01, 0.0]
m: 0.0685088 +/- 0.000416164 (99.7% conf)
c: -1.68779e-06 +/- 3.95437e-06 (99.7% conf)
```

compared to galsim shearing in `ngmix/metacal.py`:
```
S/N: 84713.4
R11: 0.401513
m: 7.92014e-06 +/- 0.000347655 (99.7% conf)
c: -9.48739e-08 +/- 3.49117e-06 (99.7% conf)
```

So the shearing also induces an error.

To summarize, this tests tell us that we need to:
- Create a more general convolution function to:
  -  enble odd stamp sizes
  - resample images on finer grids (e.g. current convolution works worse on exponentials than on Gaussians)
- Correct the shearing transform
  - change the interpolation kernel
  - check that the gradients are also correct

What I propose is to handle this points in separated issues and pull requests.

**Comment by b-remy** [comment]
The aim of this pull request was to build [utils functions](https://github.com/b-remy/gems/blob/metacal/utils.py) to handle data and metacalibration scripts that can take it as input. The tests above were made under this configuration.

So I guess this pull request can be merged and do the following in separated issues.

**Comment by EiffL** [comment]
hummmmmmmm two comments:
- We should build automated unit tests for this!!!

- Do you test the shear independently from the odd/even stamp sizes? Because the gradients of the shear are not necessarily 100% correct, but just like in Andre's results, the forward pass should be correct, and you should be able to recover an unbiased shear under finite diff metacal

**Comment by EiffL** [comment]
You can set the unit tests up in pretty much the same way as in galflow, with github actions, so that they will run automatically on PRs.
